examples/state_machine/states.py

```diff
@@ -0,0 +1,517 @@
+        )
+
+    def reconnecting(self, event: events.HandshakeFailureEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        self._context.handshake_count = event.handshake_count + 1
+        self._context.error = event.error
+
+        return PNTransition(
+            state=HandshakeReconnectingState,
+            context=self._context
+        )
+
+    def disconnect(self, event: events.DisconnectEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        self._context.tt = 0
+
+        return PNTransition(
+            state=HandshakeStoppedState,
+            context=self._context
+        )
+
+    def handshaking_success(self, event: events.HandshakeSuccessEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        return PNTransition(
+            state=ReceivingState,
+            context=self._context,
+            status=PNStatus(category=PNStatusCategory.PNConnectedCategory)
+        )
+
+
+class HandshakeReconnectingState(PNState):
+    def __init__(self, context: PNContext) -> None:
+        super().__init__(context)
+        self._transitions = {
+            events.DisconnectEvent.__name__: self.disconnect,
+            events.HandshakeReconnectGiveupEvent.__name__: self.give_up,
+            events.HandshakeReconnectSuccessEvent.__name__: self.success,
+            events.SubscriptionRestoredEvent.__name__: self.restored,
+            events.HandshakeReconnectFailureEvent.__name__: self.handshake_reconnect,
+            events.SubscriptionChangedEvent.__name__: self.subscription_changed,
+        }
+
+    def on_enter(self, context: Union[None, PNContext]):
+        self._context.update(context)
+        super().on_enter(self._context)
+        return effects.HandshakeReconnectEffect(self._context.channels, self._context.groups)
+
+    def on_exit(self):
+        super().on_exit()
+        return effects.CancelHandshakeEffect()
+
+    def disconnect(self, event: events.DisconnectEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+
+        return PNTransition(
+            state=HandshakeStoppedState,
+            context=self._context,
+            effect=effects.CancelHandshakeEffect()
+        )
+
+    def subscription_changed(self, event: events.SubscriptionChangedEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        self._context.channels = event.channels
+        self._context.groups = event.groups
+
+        return PNTransition(
+            state=HandshakeReconnectingState,
+            context=self._context
+        )
+
+    def handshake_reconnect(self, event: events.HandshakeReconnectFailureEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        self._context.handshake_count + 1
+        self._context.error = event.error
+
+        return PNTransition(
+            state=ReceivingState,
```

Comment:
In case of `HandshakeReconnectFailureEvent` I think we should try to reconnect once again, so it's not `ReceivingState` but rather `HandshakeReconnectingState`